### DISCUSSION TITLE
Setting curve X448 seems to result in setting curve X25519

### DISCUSSION THREAD

**Discussion by evilaliv3** [openssl/openssl#24116] [Q&A] [answered]
Hello,

i'm trying to enable curve X448 using the following code, but after verifying it i notice that configured curve appears to be X25519. Would you please clarify the reason and if this is correct up to your expectations or we should look into a possible regression?

The code i'm using (in python in @globaleaks) is:
```python
        # Configure ECDH with CURVE NID_x448
        ecdh = _lib.EC_KEY_new_by_curve_name(1034)
        ecdh = _ffi.gc(ecdh, _lib.EC_KEY_free)
        _lib.SSL_CTX_set_tmp_ecdh(self.ctx._context, ecdh)

```

From what i know NID_x448 is set on number 1034 and NID_x25519 on number 1035

This can be verified on host try.globaleaks.org

For example with:
- openssl s_client -connect try.globaleaks.org:443
- https://www.ssllabs.com/ssltest/analyze.html?d=try.globaleaks.org

Thank you!

**Comment by t8m** [comment, marked as answer]
This would be better discussed at the python bindings issue tracker.

However, EC_KEY_new_by_curve_name() cannot be used with X25519 and X448 curves. They use a completely separate key type.

You should use SSL_CTX_set1_groups_list() https://www.openssl.org/docs/manmaster/man3/SSL_CTX_set1_groups.html to set up the preferred groups for the key exchange.

**Comment by evilaliv3** [comment]
Thank you @t8m, this clarifies, the code was actually implemented for secp384r1.

Unfortunately the functions that you refer does not feature a binding. I suggested this in 2019 but it was not considered:
https://github.com/pyca/cryptography/issues/4980

Fortunately by enabling TLS1.2 and TLS1.3 at least the default configuration seems to be using x25519;
Do you or @alex know what causes this behaviour and if by any chance there is an alternative way to enforce the usage of only X448?

I see that at least for the versions of pyopenssl and openssl present on debian bookworm the server preference is: x25519, secp256r1, x448, secp521r1, secp384r
